src/tox/tox_env/python/virtual_env/package/pyproject.py

```diff
@@ -157,14 +159,19 @@ def _setup_env(self) -> None:
         if "editable" in self.builds:
             if not self._frontend.optional_hooks["build_editable"]:
                 raise BuildEditableNotSupportedError
-            build_requires = self._frontend.get_requires_for_build_editable().requires
-            self._install(build_requires, PythonPackageToxEnv.__name__, "requires_for_build_editable")
+            self._setup_build_requires("editable")
         if "wheel" in self.builds:
-            build_requires = self._frontend.get_requires_for_build_wheel().requires
-            self._install(build_requires, PythonPackageToxEnv.__name__, "requires_for_build_wheel")
+            self._setup_build_requires("wheel")
         if "sdist" in self.builds or "external" in self.builds:
-            build_requires = self._frontend.get_requires_for_build_sdist().requires
-            self._install(build_requires, PythonPackageToxEnv.__name__, "requires_for_build_sdist")
+            self._setup_build_requires("sdist")
+
+    def _setup_build_requires(self, of_type: str) -> None:
+        hook = getattr(self._frontend, f"get_requires_for_build_{of_type}")
+        build_requires = hook().requires  # already cached via Pep517VirtualEnvFrontend
+        pending = [req for req in build_requires if req not in self._installed_build_requirements]
```

Comment:
This filter will not work, because you'll not detect if the backend removed dependencies and fail badly 🤔